src/Microsoft.AspNetCore.Routing/CompositeEndpointDataSource.cs

```diff
@@ -0,0 +1,70 @@
+﻿// Copyright (c) .NET Foundation. All rights reserved.
+// Licensed under the Apache License, Version 2.0. See License.txt in the project root for license information.
+
+using System;
+using System.Collections.Generic;
+using System.Linq;
+using Microsoft.Extensions.Primitives;
+
+namespace Microsoft.AspNetCore.Routing
+{
+    public class CompositeEndpointDataSource : EndpointDataSource
```

Comment:
This is public so that you can get it from DI. We don't do that it in the implementation of the middleware because the initialization semantics are different.

Reply:
I think it should be real public. This is what you'd use if you want to programmatically enumerate all of the endpoints in the application.